Rework architecture document (#35)
alxdavids/privacy-pass-ietf: draft-davidson-pp-architecture.md

before being sent. We use the TLS presentation language {{RFC8446}} to
describe the structure of the data that is communicated and stored.

# Ecosystem participants {#ecosystem}

The Privacy Pass ecosystem refers to the global framework in which
multiple instances of the Privacy Pass protocol operate. This refers to
all servers that support the protocol, or any extension of it, along
with all of the Clients that may interact with these Servers.

The ecosystem itself, and the way it is constructed, is critical for
evaluating the privacy of each individual Client. We assume that a
Client's privacy refers to fraction of users that it represents in the
anonymity set that it belongs to. We discuss this more in {{privacy}}.

TODO: Add diagram of ecosystem

## Servers {#ecosystem-servers}

Generally, Servers in the Privacy Pass ecosystem are entities whose
primary function is to undertake the role of the `Issuer` in
{{draft-davidson-pp-protocol}}. To facilitate this, the Issuer MUST hold
a Privacy Pass protocol keypair at any given time. The Issuer public key
MUST be made available to all Clients in such a way that key rotations

> and other updates can be monitored. The Issuer MAY also require

Author: Chnaged to "are publicly visible"

## Clients {#ecosystem-clients}

Clients in the Privacy Pass ecosystem are entities whose primary
function is to undertake the role of the `Client` in
{{draft-davidson-pp-protocol}}. Clients are assumed to only store data
related to the tokens that it has been issued by the Server. This
storage is used for constructing redemption requests.

# Key management framework {#key-mgmt}

The key material and protocol configuration that a Server uses to issue
tokens corresponds to a number of different pieces of information.

- The ciphersuite that the Server is using.
- The public keys that are active for the Server.

For reasons that we address later in {{privacy}}, the way that the
Server publishes and maintains this information impacts the effective
privacy of the clients. In this section, we describe the main policies
that need to be satisfied for a key management system that serves a
particular Privacy Pass ecosystem.

Note that we only specify a set of guidelines and recommendations for
operating a key registry in this document. Actual specification of such
a registry and how it operates will be covered in separate
documentation.

## Public key registries

> Issuer's must provide their public keys to clients along with details

Author: Yeah I agree, I'm not sure what the best path here is though. Perhaps we should just go with Server everywhere?

This is to protect against Servers that try and track users by issuing
individual keys for each user.

We RECOMMEND that any key registry is append-only, and publishes the
timings of all Server updates. The key registry should be operated
independently of any Issuer that publishes key material to the registry.
This ensures that any Client can make better judgements on whether to
trust the registry (and each Server).

## Key rotation

Token issuance associates all issued tokens with a particular choice of
key. If a Server issues tokens with many keys, then this may harm the
anonymity of the Client. For example, they would be able to map the
Client's access patterns by inspecting which key each token they possess
has been issued under.

To prevent against this, Servers MUST only use one private key for
issuing tokens at any given time. Two keys that are used for redemption
are permitted to allow Servers to rotate keys smoothly.

Key rotations must be limited for similar reasons. See
{{parametrization}} for guidelines on what frequency of key rotations
are permitted.

## Ciphersuites

Since a Server is only permitted to have a single active issuing key,

> this transitively implies that only a single ciphersuite is allowed. If

Reviewer: ```suggestion
this implies that only a single ciphersuite is allowed per issuance period. If
```

> a Server wishes to change their ciphersuite, they should do so during a

Reviewer: ```suggestion
a Server wishes to change their ciphersuite, they MUST do so during a
```

key rotation.

> ## Checking registry integrity

Reviewer: I think we can axe this section, as it's covered by the text above.

Out-of-band checks that establish the integrity of a key registry should
be available. For example, by publishing hashes of the current registry
contents to a globally trusted location. Moreover, regular checks should
be made to ascertain whether a Server is publishing different key
material to multiple registries (see {{split-view}}).

Clients may also choose to publicly expose registries that they do not
trust. This could be done by posting on some public bulletin board, or
via other means such as Client gossip protocols.

# Server running modes {#running-modes}

We provide an overview of some of the possible frameworks for
configuring the way that Servers run in the Privacy Pass ecosystem. In
short, Servers may be configured to provide symmetric issuance and
redemption with clients. While some Servers may be configured as proxies
that accept Privacy Pass data and send it to another Server that

> actually processes issuance and/or redemption data. Finally, we also

Reviewer: nit:

```suggestion
actually processes issuance or redemption data. Finally, we also
```

consider instances of the protocol that may permit public verification.

The intention with providing each of these running modes is to cover the
different applications that utilize variants of the Privacy Pass
protocol. We RECOMMEND that any Privacy Pass Server implementation
adheres to one of these frameworks.

## Single-Issuer Single-Verifier {#sisv}

The simplest way of considering the Privacy Pass protocol is in a
setting where the same server plays the role of issuer and verifier, we
call this "Single-Issuer Single-Verifier" (SISV).

When S wants to issue tokens to C, they invoke the issuance protocol
where C generates their own inputs, and S uses their secret key skS. In
this setting, C can only perform token redemption with S. When a token
redemption is required, C and S invoke the redemption phase of the
protocol, where C uses an issued token from a previous exchange, and S
uses skS to validate the redemption.

> ## Single-Issuer Forwarding-Verifier {#sifv}

Reviewer: nit: Proxied-Verifier, maybe?

Author: Small conflict with abbreviation of public verifier. Would "Delegated-Verifier" work?

    |      +-----+        +-------------> |     |         |
    |                     |               +-----+         |
    |                     |                               |
    |      +-----+        |                               |
    |      |     | <------+                               |
    |      | C2  |                                        |
    |      |     | <------+               +-----+         |
    |      +-----+        +-------------> |     |         |
    |                                     | S2  |         |
    |                     +-------------> |     |         |
    |      +-----+        |               +-----+         |
    |      |     |        |                               |
    |      | C3  | <------+                               |
    |      |     |                                        |
    |      +-----+                                        |
    |                                                     |
    +-----------------------------------------------------+
~~~

In the above diagram, the arrows indicate the open channels between a
Client and a Server. An open channel indicates that a Client accepts

> Privacy Pass tokens from this Server. The channel also infers which
> Servers, the client can be linked to by redemption of tokens.

Reviewer: ```suggestion
Privacy Pass tokens from this Server. 
```

This ensures that any Client can make better judgements on whether to
trust the registry and, transitively, each Server.

> ~~~
> +---------------------------------------------------------------+
> |                                                               |
> | Ecosystem                                                     |
> |                                pkS1                           |
> |  +---------------------+                    +-+               |
> |  |  key_registry_1     | <----------------->|C|               |
> |  |                     |                    +-+               |
> |  ++--------------------+ <------+     pkS1                    |
> |   ^                             +---------------------> +-+   |
> |   |   +---------------------+         pkS3              |C|   |
> |   |   |  key_registry_2     | <-----------------------> +-+   |
> |   |   |                     |                                 |
> |   |   +------+---------+----+ <---------+                     |
> |   |          ^         ^          pkS2  |                     |
> |   | pkS1     | pkS2    | pkS3           |   +-+               |
> |   |          |         |                +-> |C|               |
> |  ++---+    +-+--+    +-+--+                 +-+               |
> |  |    |    |    |    |    |                                   |
> |  | S1 |    | S2 |    | S3 |                                   |
> |  |    |    |    |    |    |                                   |
> |  +----+    +----+    +----+                                   |
> |                                                               |
> |                                                               |
> +---------------------------------------------------------------+
> ~~~

Reviewer: ```suggestion
~~~
+--------------------------------------------------------+
|                                                        |
| Ecosystem                            +---+             |
|                                      | C |             |
|  +--------------+ <------ pkS1 ----> +---+             |
|  |  Registry 1  |                                      |
|  ++-------------+ <-------------- pkS1 --------> +---+ |
|   |                                              | C | |
|   |   +--------------+ <--------- pkS3 --------> +---+ |
|   |   |  Registry 2  |                                 |
|  pkS1 +----^-------^-+ <--------- pkS2 --------> +---+ |
|   |        |       |                             | C | |
|   |       pkS2    pkS3                           +---+ |
|   |        |       |                                   |
|  ++---+  +-+--+  +-+--+                                |
|  | S1 |  | S2 |  | S3 |                                |
|  +----+  +----+  +----+                                |
|                                                        |
+--------------------------------------------------------+
~~~
```

Issuers.

In addition, we RECOMMEND that trusted registries indicate at all times
which Issuers are deemed to be active. If a client is asked to invoke
any Privacy Pass exchange for an issuer that is not declared active,
then the client SHOULD refuse to retrieve the server configuration
during the protocol.

### Allowing larger number of Issuers {#more-issuers}

The bounds on the numbers of Issuers that we proposed above are very
restrictive. This is due to the fact that we considered a situation
where a Client could be issued (and forced to redeem) tokens for any
issuing key.

An alternative system is to ensure a robust strategy for ensuring that
Clients only possess redemption tokens for a similarly small number of
Issuers at any one time. This prevents a malicious verifier from being
able to invoke redemptions for many Issuers since the Client would only
be holding redemption tokens for a small set of Issuers. When a Client
is issued tokens from a new issuer and already has tokens from the
maximum number of Issuers, it simply deletes the oldest set of
redemption tokens in storage and then stores the newly acquired tokens.

> For example, if Clients ensure that they only hold redemption tokens 4

Reviewer: nit: bug in this sentence? "that they only hold redemption tokens 4 Issuers"

old key.

## Token exhaustion

When a Client holds tokens for an issuer, it is possible for any
verifier to invoke that Client to redeem tokens for that issuer. This
can lead to an attack where a malicious verifier can force a Client to
spend all of their tokens for a given issuer. To prevent this from
happening, methods should be put into place to prevent many tokens from
being redeemed at once.

For example, it may be possible to cache a redemption for the entity
that is invoking a token redemption. If the verifier requests more
tokens then the Client simply returns the cached token that it returned
previously. This could also be handled by simply not redeeming any
tokens for verification if a redemption had already occurred in a given
time window.

In SIAV, the Client instead caches the SRR that it received in the
asynchronous redemption exchange with the issuer. If the same verifier
attempts another redemption request, then the Client simply returns the
cached SRR. The SRRs can be revoked by the issuer, if need be, by
providing an expiry date or by signaling that records from a particular
window need to be refreshed.

## Avoiding Issuer centralization

> TODO: explain potential and mitigations for issue centralization

Reviewer: ```suggestion
[[OPEN ISSUE: explain potential and mitigations for Issuer centralization]]
```
